src/aatoolbox/utils/raster.py

```diff
@@ -0,0 +1,447 @@
+"""
+Utilities to manipulate and analyze raster data.
+
+This module extends xarray and rioxarray to provide
+additional functionality for raster processing and
+post-processing. The extension is based on the
+guidance for how to extend xarray:
+
+http://xarray.pydata.org/en/stable/internals/extending-xarray.html
+
+However, since rioxarray already extends xarray, this
+modules extensions inherit from the RasterArray and
+RasterDataset extensions respectively. This ensures
+cleaner code in the module as ``rio`` methods are
+available immediately, but also means a couple of
+design decisions are followed.
+
+The xarray.DataArray and xarray.DataSet
+extensions here inherit from rioxarray base classes.
+Thus, methods that are identical for both objects
+are defined in a mixin class ``AatRasterMixin`` which
+can be inherited by the two respective extensions.
+"""
+
+import logging
+from typing import Any, Callable, Dict, List, Union
+
+import geopandas as gpd
+import numpy as np
+import pandas
+import pandas as pd
+import rioxarray
+import xarray
+import xarray as xr
+from rioxarray.raster_array import RasterArray
+from rioxarray.rioxarray import _get_data_var_message
+
+logger = logging.getLogger(__name__)
+
+
+class AatRasterMixin:
+    """AA toolbox mixin base class."""
+
+    # setting attributes to avoid mypy error, from SO
+    # https://stackoverflow.com/questions/53120262/mypy-how-to-
+    # ignore-missing-attribute-errors-in-mixins/53228204#53228204
+    _get_obj: Callable
+
+    def __init__(self, xarray_obj):
+        super().__init__(xarray_obj)
+
+        # Adding lat/lon to set of default spatial dims
+        if "lat" in self._obj.dims and "lon" in self._obj.dims:
+            self._x_dim = "lon"
+            self._y_dim = "lat"
+
+        # Managing time coordinate default dims
+        self._t_dim = None
+        for t in ["t", "T"]:
+            if t in self._obj.dims:
+                self._t_dim = t
+
+    # methods derived from rioxarray.rioxarray.x_dim and y_dim
+    @property
+    def t_dim(self):
+        """str: The dimension for time."""
+        if self._t_dim is not None:
+            return self._t_dim
+        raise rioxarray.exceptions.DimensionError(
+            "Time dimension not found. 'aat.set_time_dim()' or "
+            "using 'rename()' to change the dimension name to "
+            f"'t' can address this.{_get_data_var_message(self._obj)}"
+        )
+
+    def set_time_dim(
+        self, t_dim: str, inplace: bool = False
+    ) -> Union[xarray.DataArray, xarray.DataSet]:
```

Comment:
Make sure to use `Dataset` instead of `DataSet` (it's in a few more places as well)